src/sinks/elasticsearch/mod.rs

```diff
@@ -494,59 +497,61 @@ enum ParseError {
 }
 
 impl ElasticSearchCommon {
-    fn build_action(
-        &self,
-        index: &str,
-        bulk_action: BulkAction,
-        event: &mut Event,
-    ) -> serde_json::Value {
+    fn encode_log(&self, event: Event) -> Option<EncodedEvent<Vec<u8>>> {
+        let index = self.mode.index(&event)?;
+
+        let mut event = if let Some(cfg) = self.mode.as_data_stream_config() {
+            cfg.update_timestamp(cfg.sync_fields(event))
+        } else {
+            event
+        };
+
+        let bulk_action = self.mode.bulk_action(&event)?;
+
         let mut action = json!({
             bulk_action.as_str(): {
                 "_index": index,
                 "_type": self.doc_type,
             }
         });
+
         maybe_set_id(
             self.id_key.as_ref(),
             action.pointer_mut(bulk_action.as_json_pointer()).unwrap(),
-            event,
+            &mut event,
         );
-        action
-    }
-}
-
-#[async_trait::async_trait]
-impl HttpSink for ElasticSearchCommon {
-    type Input = Vec<u8>;
-    type Output = Vec<u8>;
-
-    fn encode_event(&self, event: Event) -> Option<EncodedEvent<Self::Input>> {
-        let index = self.mode.index(&event)?;
-
-        let mut event = if let Some(cfg) = self.mode.as_data_stream_config() {
-            cfg.update_timestamp(cfg.sync_fields(event))
-        } else {
-            event
-        };
-
-        let bulk_action = self.mode.bulk_action(&event)?;
-        let action = self.build_action(&index, bulk_action, &mut event);
 
         let mut body = serde_json::to_vec(&action).unwrap();
         body.push(b'\n');
 
         self.encoding.apply_rules(&mut event);
 
-        serde_json::to_writer(&mut body, &event.into_log()).unwrap();
+        serde_json::to_writer(&mut body, event.as_log()).unwrap();
         body.push(b'\n');
 
         emit!(ElasticSearchEventEncoded {
             byte_size: body.len(),
-            index
+            index,
         });
 
         Some(EncodedEvent::new(body))
     }
+}
+
+#[async_trait::async_trait]
+impl HttpSink for ElasticSearchCommon {
+    type Input = Vec<u8>;
+    type Output = Vec<u8>;
+
+    fn encode_event(&self, event: Event) -> Option<EncodedEvent<Self::Input>> {
+        match event {
+            Event::Log(_) => self.encode_log(event),
+            Event::Metric(metric) => self
+                .metric_to_log
+                .transform_one(metric)
+                .and_then(|log| self.encode_log(log.into())),
+        }
```

Comment:
I suggest converting down to log and then encoding:
```suggestion
        let log = match event {
            Event::Log(_) => Some(event),
            Event::Metric(metric) => self.metric_to_log.transform_one(metric),
        };
        log.and_then(|log| self.encode_log(log.into())
```